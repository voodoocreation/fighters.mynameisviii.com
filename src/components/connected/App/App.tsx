import { NextContext } from "next";
import withReduxSaga from "next-redux-saga";
import withRedux from "next-redux-wrapper";
import { DefaultQuery } from "next-server/router";
import NextApp, {
  AppProps,
  Container,
  DefaultAppIProps,
  NextAppContext
} from "next/app";
import * as React from "react";
import { addLocaleData } from "react-intl";
import { IntlProvider } from "react-intl-redux";
import { Provider } from "react-redux";

import routes from "../../../../next.routes";
import * as actions from "../../../actions/root.actions";
import { isServer } from "../../../helpers/dom";
import { createStore, TStore } from "../../../store/root.store";
import Page from "../Page/Page";

import en from "react-intl/locale-data/en";

import "../../../scss/index.scss";

addLocaleData([...en]);

export type TContext<Q extends DefaultQuery> = NextContext<Q> & {
  store: TStore;
};

export type TAppContext<Q extends DefaultQuery> = NextAppContext & {
  ctx: TContext<Q>;
  store: TStore;
};

interface IProps extends DefaultAppIProps, AppProps {
  intlProps: {
    locale: string;
    initialNow: Date;
  };
  store: TStore;
}

const getIntlProps = (ctx: NextContext) => {
  const requestProps = isServer()
    ? ctx.req
    : window.__NEXT_DATA__.props.initialProps.intlProps;
  const { locale } = requestProps;

  return {
    initialNow: Date.now(),
    locale: locale || "en-NZ"
  };
};

class App<P extends IProps> extends NextApp<P> {
  public static async getInitialProps({ ctx, Component }: TAppContext<any>) {
    let pageProps = {};

    const unsubscribe = ctx.store.subscribe(() => {
      const state = ctx.store.getState();

      if (state.app.error && ctx.res) {
        ctx.res.statusCode = state.app.error.status;
        unsubscribe();
      }
    });

    if (Component.getInitialProps) {
      pageProps = await Component.getInitialProps(ctx);
    }

    const intlProps = getIntlProps(ctx);

    return { pageProps, intlProps };
  }

  public readonly serviceWorkerContainer: any = undefined;

  constructor(props: P) {
    super(props);

    if (
      !isServer() &&
      "serviceWorker" in navigator &&
      (window.location.protocol === "https:" ||
        window.location.hostname === "localhost")
    ) {
      // tslint:disable-next-line
      this.serviceWorkerContainer = require("serviceworker-webpack-plugin/lib/runtime");
    }

    routes.Router.onRouteChangeStart = this.onRouteChangeStart;
    routes.Router.onRouteChangeComplete = this.onRouteChangeComplete;
    routes.Router.onRouteChangeError = this.onRouteChangeError;
  }

  public componentDidMount() {
    const { store } = this.props;

    store.dispatch(actions.setCurrentRoute(routes.Router.route));

    if (!isServer()) {
      window.addEventListener("feature", this.onReceiveFeature);

      store.dispatch(
        actions.setIsInInstalledApp(
          window.matchMedia("(display-mode: standalone)").matches
        )
      );
    }

    if (this.serviceWorkerContainer) {
      navigator.serviceWorker.addEventListener(
        "message",
        this.onReceiveServiceWorkerPostMessage
      );

      window.addEventListener(
        "beforeinstallprompt",
        this.onBeforeInstallPrompt
      );

      this.serviceWorkerContainer.register({ scope: "/" });
    }
  }

  public componentWillUnmount() {
    window.removeEventListener("feature", this.onReceiveFeature);

    if (this.serviceWorkerContainer) {
      navigator.serviceWorker.removeEventListener(
        "message",
        this.onReceiveServiceWorkerPostMessage
      );

      window.removeEventListener(
        "beforeinstallprompt",
        this.onBeforeInstallPrompt
      );
    }
  }

  public render() {
    const { Component, intlProps, pageProps, store } = this.props;

    return (
      <Container>
        <Provider store={store}>
          <IntlProvider
            defaultLocale="en-NZ"
            initialNow={intlProps.initialNow}
            textComponent={React.Fragment}
          >
            <Page>
              <Component {...pageProps} />
            </Page>
          </IntlProvider>
        </Provider>
      </Container>
    );
  }

  private onReceiveFeature = (event: any) => {
    const { store } = this.props as IProps;

    store.dispatch(actions.addFeatures(event.detail));
  };

  private onReceiveServiceWorkerPostMessage = (event: any) => {
    const { store } = this.props;

    store.dispatch(actions.receiveServiceWorkerMessage(event.data));
  };

  private onBeforeInstallPrompt = async (event: any) => {
    const { store } = this.props;
    const { outcome } = await event.userChoice;

    store.dispatch(
      actions.trackEvent({
        event: "page.addToHomeScreen.outcome",
        value: outcome
      })
    );
  };

  private onRouteChangeStart = (path: string) => {
    const { store } = this.props;

    store.dispatch(actions.changeRoute.started(path));
  };

  private onRouteChangeComplete = (path: string) => {
    const { store } = this.props;

    store.dispatch(actions.changeRoute.done({ params: path, result: {} }));
  };

  private onRouteChangeError = (error: Error, path: string) => {
    const { store } = this.props;

    store.dispatch(
      actions.changeRoute.failed({
        error: {
          message: error.toString(),
          status: 500
        },
        params: path
      })
    );
  };
}

const AppWrapped = withRedux(createStore)(withReduxSaga(App));

export default AppWrapped;
