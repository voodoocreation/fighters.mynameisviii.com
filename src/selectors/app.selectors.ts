import { IStoreState } from "../reducers/root.reducers";

export const getCurrentRoute = (state: IStoreState) => state.app.currentRoute;

export const getAppError = (state: IStoreState) => state.app.error;

export const isLoadingApp = (state: IStoreState) => state.app.isLoading;

export const hasNewVersion = (state: IStoreState) => state.app.hasNewVersion;

export const isInInstalledApp = (state: IStoreState) =>
  state.app.isInInstalledApp;
