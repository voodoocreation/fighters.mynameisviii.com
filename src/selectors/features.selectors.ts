import { IStoreState } from "../reducers/root.reducers";

export const getFeatures = (state: IStoreState) => state.features.items;

export const hasFeature = (state: IStoreState, feature: string) =>
  getFeatures(state).includes(feature);
