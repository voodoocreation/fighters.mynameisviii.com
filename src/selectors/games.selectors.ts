import { createSelector } from "reselect";

import { IStoreState } from "../reducers/root.reducers";

export const getGames = (state: IStoreState) => state.games.items;

export const getCurrentGameSlug = (state: IStoreState) =>
  state.games.currentSlug;

export const getCurrentGame = createSelector(
  getGames,
  getCurrentGameSlug,
  (games, slug) => games.find(game => game.slug === slug)
);
