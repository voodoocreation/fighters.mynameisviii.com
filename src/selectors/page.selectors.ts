import { IRootReducers } from "../reducers/root.reducers";

export const getCurrentRoute = (state: IRootReducers) =>
  state.page.currentRoute;

export const getPageError = (state: IRootReducers) => state.page.error;

export const getPageIsLoading = (state: IRootReducers) => state.page.isLoading;

export const hasNewVersion = (state: IRootReducers) => state.page.hasNewVersion;

export const isInInstalledApp = (state: IRootReducers) =>
  state.page.isInInstalledApp;
