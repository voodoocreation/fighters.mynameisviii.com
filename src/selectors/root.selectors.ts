export * from "./app.selectors";
export * from "./features.selectors";
export * from "./games.selectors";
