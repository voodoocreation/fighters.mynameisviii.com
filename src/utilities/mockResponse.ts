import { failure, success } from "../models/root.models";

export const mockWithData = <T>(data: T) => jest.fn(() => success(data));

export const mockWithError = (message: string) =>
  jest.fn(() => failure(message));

export const mockWithPayload = () =>
  jest.fn(<P>(payload: P) => success(payload));
