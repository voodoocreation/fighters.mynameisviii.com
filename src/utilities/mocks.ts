import { failure, success } from "../models/root.models";

export const createMockElement = (
  width = 0,
  height = 0,
  top = 0,
  left = 0
) => ({
  getBoundingClientRect: () => ({
    bottom: top + height,
    left,
    right: left + width,
    top
  })
});

export const findMockCall = (mockFn: any, ...args: any[]) =>
  mockFn.mock.calls.find((call: any) =>
    args.reduce((acc, curr, index) => acc && call[index] === curr, true)
  );

export const mockWithData = <T>(data: T) => jest.fn(() => success(data));

export const mockWithError = (message: string) =>
  jest.fn(() => failure(message));

export const mockWithPayload = () =>
  jest.fn(<P>(payload: P) => success(payload));
