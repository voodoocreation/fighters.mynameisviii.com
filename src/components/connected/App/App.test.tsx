import { mount } from "enzyme";
import merge from "lodash.merge";
import * as React from "react";

import { findMockCall } from "../../../utilities/mocks";

import App from "./App";

jest.mock("serviceworker-webpack-plugin/lib/runtime", () => ({
  register: jest.fn(() => new Promise(resolve => resolve()))
}));

jest.mock("../../../../next.routes", () => ({
  Router: {
    route: ""
  }
}));

import routes from "../../../../next.routes";

import * as selectors from "../../../selectors/root.selectors";

const g: any = global;

const setup = async (
  fn: any,
  fromTestProps?: any,
  isServer = false,
  isInInstalledApp = false
) => {
  jest.clearAllMocks();

  g.matchMedia = jest.fn(() => ({ matches: isInInstalledApp }));

  g.isServer = isServer;
  g.__NEXT_DATA__ = {
    props: {
      initialProps: {
        intlProps: {}
      }
    }
  };

  const Component = () => <div className="PageComponent" />;
  const appProps = merge(
    {
      Component,
      asPath: "",
      ctx: {
        isServer: false,
        pathname: "",
        query: {},
        req: {
          intlMessages: {},
          locale: "en-NZ"
        },
        res: {}
      },
      router: {
        pathname: ""
      }
    },
    fromTestProps
  );
  const initialProps = await App.getInitialProps(appProps);
  const props = {
    ...appProps,
    ...initialProps
  };

  return {
    props,
    wrapper: fn(<App {...props} />)
  };
};

describe("[connected] <App />", () => {
  const addEventListener = g.addEventListener;
  const removeEventListener = g.removeEventListener;

  beforeAll(() => {
    g.addEventListener = jest.fn((...args) => addEventListener(...args));
    g.removeEventListener = jest.fn((...args) => removeEventListener(...args));
  });

  it("mounts application correctly on the server", async () => {
    const { wrapper } = await setup(
      mount,
      {
        ctx: { isServer: g.isServer }
      },
      true
    );

    expect(wrapper.render()).toMatchSnapshot();
    wrapper.unmount();
  });

  it("mounts application correctly on the client", async () => {
    const { wrapper } = await setup(mount);

    expect(wrapper.render()).toMatchSnapshot();
    wrapper.unmount();
  });

  it("gets `initialProps` from component correctly", async () => {
    const test = "Test";
    const Component: any = () => <div className="PageComponent" />;

    Component.getInitialProps = async () => ({ test });

    const { props, wrapper } = await setup(mount, { Component });

    expect(props.initialProps.pageProps).toEqual({ test });

    wrapper.unmount();
  });

  describe("when checking if the user is in the installed app", () => {
    describe("when the display mode is 'standalone'", () => {
      let result: any;

      it("mounts the component", async () => {
        result = await setup(mount, {}, false, true);
      });

      it("makes a call to window.matchMedia", () => {
        expect(g.matchMedia).toHaveBeenCalledTimes(1);
      });

      it("isInInstalledApp is true in the store", () => {
        expect(
          selectors.isInInstalledApp(result.props.ctx.store.getState())
        ).toBe(true);
      });

      it("unmounts the component", () => {
        result.wrapper.unmount();
      });
    });

    describe("when the display mode isn't 'standalone'", () => {
      let result: any;

      it("mounts the component", async () => {
        result = await setup(mount, {}, false, false);
      });

      it("makes a call to window.matchMedia", () => {
        expect(g.matchMedia).toHaveBeenCalledTimes(1);
      });

      it("isInInstalledApp is false in the store", () => {
        expect(
          selectors.isInInstalledApp(result.props.ctx.store.getState())
        ).toBe(false);
      });

      it("unmounts the component", () => {
        result.wrapper.unmount();
      });
    });
  });

  describe("when receiving features from the feature event", () => {
    const feature1 = "feature-1";
    const feature2 = "feature-2";
    let result: any;

    it("mounts the component", async () => {
      result = await setup(mount);
    });

    it("receives a single feature correctly", async () => {
      window.dispatchEvent(new CustomEvent("feature", { detail: feature1 }));

      expect(
        selectors.hasFeature(result.props.ctx.store.getState(), feature1)
      ).toBe(true);
    });

    it("receives multiple features correctly", async () => {
      window.dispatchEvent(
        new CustomEvent("feature", {
          detail: [feature1, feature2]
        })
      );

      expect(
        selectors.hasFeature(result.props.ctx.store.getState(), feature1)
      ).toBe(true);
      expect(
        selectors.hasFeature(result.props.ctx.store.getState(), feature2)
      ).toBe(true);
    });

    it("unmounts the component", () => {
      result.wrapper.unmount();
    });
  });

  describe("when navigating within the app", () => {
    let result: any;

    it("mounts the component", async () => {
      result = await setup(mount);
    });

    it("handles the onRouteChangeStart event correctly", async () => {
      routes.Router.onRouteChangeStart("/");

      expect(result.props.ctx.store.getState().app.transitioningTo).toBe("/");
    });

    it("handles the onRouteChangeComplete event correctly", async () => {
      routes.Router.onRouteChangeComplete("/");

      expect(
        result.props.ctx.store.getState().app.transitioningTo
      ).toBeUndefined();
      expect(result.props.ctx.store.getState().app.currentRoute).toBe("/");
    });

    it("handles the onRouteChangeError event correctly", async () => {
      routes.Router.onRouteChangeError(new Error("Server error"), "/");

      expect(
        result.props.ctx.store.getState().app.transitioningTo
      ).toBeUndefined();
      expect(result.props.ctx.store.getState().app.error).toEqual({
        message: "Error: Server error",
        status: 500
      });
    });

    it("unmounts the component", () => {
      result.wrapper.unmount();
    });
  });

  describe("when the browser prompts the user to install the app", () => {
    it("handles dismissed outcome correctly", async () => {
      let isPassing = true;

      try {
        const { wrapper } = await setup(mount);

        const event: any = new Event("beforeinstallprompt");
        event.userChoice = new Promise(resolve =>
          resolve({ outcome: "dismissed" })
        );
        await window.dispatchEvent(event);

        wrapper.unmount();
      } catch (error) {
        isPassing = false;
      }

      expect(isPassing).toBe(true);
    });

    it("handles accepted outcome correctly", async () => {
      let isPassing = true;

      try {
        const { wrapper } = await setup(mount);

        const event: any = new Event("beforeinstallprompt");
        event.userChoice = new Promise(resolve =>
          resolve({ outcome: "accepted" })
        );
        await window.dispatchEvent(event);

        wrapper.unmount();
      } catch (error) {
        isPassing = false;
      }

      expect(isPassing).toBe(true);
    });
  });

  describe("when registering the service worker", () => {
    let result: any;

    it("mounts the component", async () => {
      result = await setup(mount);
    });

    it("creates an instance of the service worker and registers it", () => {
      const instance = result.wrapper
        .childAt(0)
        .childAt(0)
        .instance();

      expect(instance.serviceWorkerContainer).toBeDefined();
      expect(instance.serviceWorkerContainer.register).toHaveBeenCalledWith({
        scope: "/"
      });
    });

    it("binds the 'message' event listener to the service worker", () => {
      expect(
        findMockCall(navigator.serviceWorker.addEventListener, "message")
      ).toBeDefined();
    });

    it("receives messages from service worker correctly", () => {
      navigator.serviceWorker.dispatchEvent(
        new MessageEvent("message", {
          data: { type: "serviceWorker.activate" }
        })
      );

      expect(result.props.ctx.store.getState().app.hasNewVersion).toBe(true);
    });

    it("unmounts the component", () => {
      result.wrapper.unmount();
    });

    it("unbinds the 'message' event listener from the service worker", () => {
      expect(
        findMockCall(navigator.serviceWorker.removeEventListener, "message")
      ).toBeDefined();
    });

    it("doesn't receive any messages after the component unmounts", () => {
      navigator.serviceWorker.dispatchEvent(
        new MessageEvent("message", {
          data: { type: "serviceWorker.activate" }
        })
      );

      expect(result.props.ctx.store.getState().app.hasNewVersion).toBe(true);
    });
  });
});
