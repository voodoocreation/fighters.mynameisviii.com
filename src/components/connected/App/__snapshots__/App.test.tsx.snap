// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[connected] <App /> mounts application correctly on the client 1`] = `
<article
  class="Page"
>
  <nav
    class="Navigation"
    role="navigation"
  >
    <section
      class="Navigation--primary"
    >
      <button
        class="Button Navigation--primary--games"
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 640 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M480 96H160C71.6 96 0 167.6 0 256s71.6 160 160 160c44.8 0 85.2-18.4 114.2-48h91.5c29 29.6 69.5 48 114.2 48 88.4 0 160-71.6 160-160S568.4 96 480 96zM256 276c0 6.6-5.4 12-12 12h-52v52c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-52H76c-6.6 0-12-5.4-12-12v-40c0-6.6 5.4-12 12-12h52v-52c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h52c6.6 0 12 5.4 12 12v40zm184 68c-26.5 0-48-21.5-48-48s21.5-48 48-48 48 21.5 48 48-21.5 48-48 48zm80-80c-26.5 0-48-21.5-48-48s21.5-48 48-48 48 21.5 48 48-21.5 48-48 48z"
          />
        </svg>
         
        <span>
          Games
        </span>
      </button>
      <a
        class="Navigation--primary--home"
        href="/"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M280.37 148.26L96 300.11V464a16 16 0 0 0 16 16l112.06-.29a16 16 0 0 0 15.92-16V368a16 16 0 0 1 16-16h64a16 16 0 0 1 16 16v95.64a16 16 0 0 0 16 16.05L464 480a16 16 0 0 0 16-16V300L295.67 148.26a12.19 12.19 0 0 0-15.3 0zM571.6 251.47L488 182.56V44.05a12 12 0 0 0-12-12h-56a12 12 0 0 0-12 12v72.61L318.47 43a48 48 0 0 0-61 0L4.34 251.47a12 12 0 0 0-1.6 16.9l25.5 31A12 12 0 0 0 45.15 301l235.22-193.74a12.19 12.19 0 0 1 15.3 0L530.9 301a12 12 0 0 0 16.9-1.6l25.5-31a12 12 0 0 0-1.7-16.93z"
          />
        </svg>
         
        <span>
          Home
        </span>
      </a>
      <button
        class="Button Navigation--primary--characters"
        disabled=""
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 640 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M96 224c35.3 0 64-28.7 64-64s-28.7-64-64-64-64 28.7-64 64 28.7 64 64 64zm448 0c35.3 0 64-28.7 64-64s-28.7-64-64-64-64 28.7-64 64 28.7 64 64 64zm32 32h-64c-17.6 0-33.5 7.1-45.1 18.6 40.3 22.1 68.9 62 75.1 109.4h66c17.7 0 32-14.3 32-32v-32c0-35.3-28.7-64-64-64zm-256 0c61.9 0 112-50.1 112-112S381.9 32 320 32 208 82.1 208 144s50.1 112 112 112zm76.8 32h-8.3c-20.8 10-43.9 16-68.5 16s-47.6-6-68.5-16h-8.3C179.6 288 128 339.6 128 403.2V432c0 26.5 21.5 48 48 48h288c26.5 0 48-21.5 48-48v-28.8c0-63.6-51.6-115.2-115.2-115.2zm-223.7-13.4C161.5 263.1 145.6 256 128 256H64c-35.3 0-64 28.7-64 64v32c0 17.7 14.3 32 32 32h65.9c6.3-47.4 34.9-87.3 75.2-109.4z"
          />
        </svg>
         
        <span>
          Characters
        </span>
      </button>
    </section>
    <section
      class="Navigation--games"
    >
      <h2>
        Games
      </h2>
      <div
        class="Navigation--games--items"
      >
        <article
          class="GameListing"
        >
          <a
            href="/games/mortal-kombat"
          >
            <img
              alt="View all moves for Mortal Kombat"
              src="/static/games/mortal-kombat/thumbnail.jpg"
            />
          </a>
        </article>
        <article
          class="GameListing"
        >
          <a
            href="/games/mortal-kombat-ii"
          >
            <img
              alt="View all moves for Mortal Kombat II"
              src="/static/games/mortal-kombat-ii/thumbnail.jpg"
            />
          </a>
        </article>
      </div>
    </section>
  </nav>
  <main
    class="Page--body"
    role="main"
  >
    <div
      class="PageComponent"
    />
  </main>
  <section
    class="ToastContainer"
  />
</article>
`;

exports[`[connected] <App /> mounts application correctly on the server 1`] = `
<article
  class="Page"
>
  <nav
    class="Navigation"
    role="navigation"
  >
    <section
      class="Navigation--primary"
    >
      <button
        class="Button Navigation--primary--games"
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 640 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M480 96H160C71.6 96 0 167.6 0 256s71.6 160 160 160c44.8 0 85.2-18.4 114.2-48h91.5c29 29.6 69.5 48 114.2 48 88.4 0 160-71.6 160-160S568.4 96 480 96zM256 276c0 6.6-5.4 12-12 12h-52v52c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-52H76c-6.6 0-12-5.4-12-12v-40c0-6.6 5.4-12 12-12h52v-52c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v52h52c6.6 0 12 5.4 12 12v40zm184 68c-26.5 0-48-21.5-48-48s21.5-48 48-48 48 21.5 48 48-21.5 48-48 48zm80-80c-26.5 0-48-21.5-48-48s21.5-48 48-48 48 21.5 48 48-21.5 48-48 48z"
          />
        </svg>
         
        <span>
          GAMES
        </span>
      </button>
      <a
        class="Navigation--primary--home"
        href="/"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M280.37 148.26L96 300.11V464a16 16 0 0 0 16 16l112.06-.29a16 16 0 0 0 15.92-16V368a16 16 0 0 1 16-16h64a16 16 0 0 1 16 16v95.64a16 16 0 0 0 16 16.05L464 480a16 16 0 0 0 16-16V300L295.67 148.26a12.19 12.19 0 0 0-15.3 0zM571.6 251.47L488 182.56V44.05a12 12 0 0 0-12-12h-56a12 12 0 0 0-12 12v72.61L318.47 43a48 48 0 0 0-61 0L4.34 251.47a12 12 0 0 0-1.6 16.9l25.5 31A12 12 0 0 0 45.15 301l235.22-193.74a12.19 12.19 0 0 1 15.3 0L530.9 301a12 12 0 0 0 16.9-1.6l25.5-31a12 12 0 0 0-1.7-16.93z"
          />
        </svg>
         
        <span>
          HOME
        </span>
      </a>
      <button
        class="Button Navigation--primary--characters"
        disabled=""
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 640 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M96 224c35.3 0 64-28.7 64-64s-28.7-64-64-64-64 28.7-64 64 28.7 64 64 64zm448 0c35.3 0 64-28.7 64-64s-28.7-64-64-64-64 28.7-64 64 28.7 64 64 64zm32 32h-64c-17.6 0-33.5 7.1-45.1 18.6 40.3 22.1 68.9 62 75.1 109.4h66c17.7 0 32-14.3 32-32v-32c0-35.3-28.7-64-64-64zm-256 0c61.9 0 112-50.1 112-112S381.9 32 320 32 208 82.1 208 144s50.1 112 112 112zm76.8 32h-8.3c-20.8 10-43.9 16-68.5 16s-47.6-6-68.5-16h-8.3C179.6 288 128 339.6 128 403.2V432c0 26.5 21.5 48 48 48h288c26.5 0 48-21.5 48-48v-28.8c0-63.6-51.6-115.2-115.2-115.2zm-223.7-13.4C161.5 263.1 145.6 256 128 256H64c-35.3 0-64 28.7-64 64v32c0 17.7 14.3 32 32 32h65.9c6.3-47.4 34.9-87.3 75.2-109.4z"
          />
        </svg>
         
        <span>
          CHARACTERS
        </span>
      </button>
    </section>
    <section
      class="Navigation--games"
    >
      <h2>
        GAMES
      </h2>
      <div
        class="Navigation--games--items"
      >
        <article
          class="GameListing"
        >
          <a
            href="/games/mortal-kombat"
          >
            <img
              alt="VIEW_ALL_MOVES_FOR_GAME"
              src="/static/games/mortal-kombat/thumbnail.jpg"
            />
          </a>
        </article>
        <article
          class="GameListing"
        >
          <a
            href="/games/mortal-kombat-ii"
          >
            <img
              alt="VIEW_ALL_MOVES_FOR_GAME"
              src="/static/games/mortal-kombat-ii/thumbnail.jpg"
            />
          </a>
        </article>
      </div>
    </section>
  </nav>
  <main
    class="Page--body"
    role="main"
  >
    <div
      class="PageComponent"
    />
  </main>
  <section
    class="ToastContainer"
  />
</article>
`;
